tests/trainer/test_scheduler.py

```diff
@@ -1,223 +1,112 @@
-        ]
-        scheduler = ComposedScheduler(schedulers)
-        self._test(scheduler, targets, epochs, optimizer)
-
-    def test_composed_linear2(self, optimizer: Optimizer):
-        epochs = 9
-        targets = [[1 * 0.5 + (x/4 * 0.5) for x in range(4)] + \
-                   [1 * 0.9**x for x in range(2)] + [1 * 0.9**x * 0.1 for x in range(2, 7)]]
-        schedulers = [
-            ExponentialLR(optimizer, gamma=0.9),
-            MultiStepLR(optimizer, milestones=[6], gamma=0.1),
-            WarmUpLR(optimizer, warmup_factor=0.5, warmup_iters=4, warmup_method="linear")
-        ]
-        scheduler = ComposedScheduler(schedulers)
-        self._test(scheduler, targets, epochs, optimizer)
-
-    def test_composed_linear_from_zero(self, optimizer: Optimizer):
-        epochs = 9
-        targets = [[1 * 0.0 + (x / 4 * 1.0) for x in range(4)] + [1 * 0.9**x for x in range(7)]]
-        schedulers = [
-            ExponentialLR(optimizer, gamma=0.9),
-            WarmUpLR(optimizer, warmup_factor=0, warmup_iters=4, warmup_method="linear")
-        ]
-        scheduler = ComposedScheduler(schedulers)
-        self._test(scheduler, targets, epochs, optimizer)
-
-    def test_composed_linear_from_zero_step(self, optimizer: Optimizer):
-        epochs = 9
-        targets = [[x / 4 for x in range(4)] + [1.0 for _ in range(7)]]
-        schedulers = [
-            ExponentialLR(optimizer, gamma=0.9),
-            WarmUpLR(optimizer, warmup_factor=0, warmup_iters=4, warmup_method="linear"),
-        ]
-        schedulers[0].interval = 'epoch'  # should never trigger
-        schedulers[1].interval = 'batch'
-        scheduler = ComposedScheduler(schedulers)
-        self._test(scheduler, targets, epochs, optimizer, interval='batch')
-
-    @pytest.mark.xfail
-    def test_validate_compose_multistep(self, optimizer: Optimizer):
-        schedulers = [
-            ExponentialLR(optimizer, gamma=0.9),
-            WarmUpLR(optimizer, warmup_factor=0, warmup_iters=4, warmup_method="linear"),
-            MultiStepLR(optimizer, milestones=[3], gamma=0.1)
-        ]
-
-        with pytest.raises(ValueError):
-            ComposedScheduler(schedulers)
-
-    @pytest.mark.xfail
-    def test_validate_compose_step(self, optimizer: Optimizer):
-        schedulers = [
-            ExponentialLR(optimizer, gamma=0.9),
-            WarmUpLR(optimizer, warmup_factor=0, warmup_iters=4, warmup_method="linear"),
-            StepLR(optimizer, step_size=2, gamma=0.1)
-        ]
-
-        with pytest.raises(ValueError):
-            ComposedScheduler(schedulers)
+def dummy_schedulers_state(dummy_model: Model, dummy_train_dataloader: DataLoader):
+    return State(
+        model=dummy_model,
+        train_dataloader=dummy_train_dataloader,
+        max_duration=MAX_DURATION,
+        steps_per_epoch=STEPS_PER_EPOCH,
+    )
+
+
+@pytest.mark.parametrize("scheduler,ssr,test_times,expected_lrs", [
```

Comment:
🤯

Reply:
We can eventually clean this up with the `pytest_generate_tests` hook, and keep the data here in a dictionary, but not required for this PR. Or adopt the pytorch approach (https://github.com/pytorch/pytorch/blob/03662b32d55b33eb6a2d63bb8ccec4671e9d323f/test/test_optim.py#L1150), which i actually find a bit too verbose.